crates/chain/src/tx_graph.rs

```diff
@@ -389,100 +385,67 @@ impl<A: Clone + Ord> TxGraph<A> {
     /// `anchor`.
-    ///
-    /// [`insert_anchor_preview`]: Self::insert_anchor_preview
-    /// [`apply_additions`]: Self::apply_additions
-    pub fn insert_anchor(&mut self, txid: Txid, anchor: A) -> Additions<A> {
-        let additions = self.insert_anchor_preview(txid, anchor);
-        self.apply_additions(additions.clone());
-        additions
+    pub fn insert_anchor(&mut self, txid: Txid, anchor: A) -> ChangeSet<A> {
+        let mut update = Self::default();
+        update.anchors.insert((anchor, txid));
+        let changeset = self.determine_changeset(&update);
+        self.apply_changeset(changeset.clone());
+        changeset
     }
 
-    /// Returns the resultant [`Additions`] if the `txid` is set to `seen_at`.
+    /// Inserts the given `seen_at` for `txid` into [`TxGraph`].
     ///
     /// Note that [`TxGraph`] only keeps track of the lastest `seen_at`.
-    pub fn insert_seen_at_preview(&self, txid: Txid, seen_at: u64) -> Additions<A> {
+    pub fn insert_seen_at(&mut self, txid: Txid, seen_at: u64) -> ChangeSet<A> {
         let mut update = Self::default();
         let (_, _, update_last_seen) = update.txs.entry(txid).or_default();
         *update_last_seen = seen_at;
-        self.determine_additions(&update)
-    }
-
-    /// Inserts the given `seen_at` into [`TxGraph`].
-    ///
-    /// This is equivalent to calling [`insert_seen_at_preview`] and [`apply_additions`] in
-    /// sequence.
-    ///
-    /// [`insert_seen_at_preview`]: Self::insert_seen_at_preview
-    /// [`apply_additions`]: Self::apply_additions
-    pub fn insert_seen_at(&mut self, txid: Txid, seen_at: u64) -> Additions<A> {
-        let additions = self.insert_seen_at_preview(txid, seen_at);
-        self.apply_additions(additions.clone());
-        additions
+        let changeset = self.determine_changeset(&update);
+        self.apply_changeset(changeset.clone());
+        changeset
     }
 
     /// Extends this graph with another so that `self` becomes the union of the two sets of
     /// transactions.
     ///
-    /// The returned [`Additions`] is the set difference between `update` and `self` (transactions that
+    /// The returned [`ChangeSet`] is the set difference between `update` and `self` (transactions that
     /// exist in `update` but not in `self`).
-    pub fn apply_update(&mut self, update: TxGraph<A>) -> Additions<A> {
-        let additions = self.determine_additions(&update);
-        self.apply_additions(additions.clone());
-        additions
+    pub fn apply_update(&mut self, update: TxGraph<A>) -> ChangeSet<A> {
+        let changeset = self.determine_changeset(&update);
+        self.apply_changeset(changeset.clone());
+        changeset
+    }
+
+    /// Determines the changeset between self and the empty [`TxGraph`].
+    pub fn initial_changeset(&self) -> ChangeSet<A> {
+        Self::default().apply_update(self.clone())
```

Comment:
Nit: This might be more efficient as we don't need to apply real changes to `Self::default()`.

```suggestion
        Self::default().determine_changeset(self)
```